src/commands/importFromRegistry.ts

```diff
@@ -0,0 +1,167 @@
+import * as vscode from "vscode";
+import { Keychain } from "../keychain";
+
+export async function importFromRegistry(scope?: vscode.ConfigurationScope) {
+  const cmd = require("node-cmd");
+  const vsWinReg = require("vscode-windows-registry");
+  const hkeyLocalMachine = "HKEY_LOCAL_MACHINE";
+
+  const config = vscode.workspace.getConfiguration("intersystems", scope);
+  const serverDefinitions: any = config.get("servers");
+
+  const newServerNames = new Array<string>();
+  const serversMissingUsernames = new Array<string>();
+  const cancellationToken = new vscode.CancellationTokenSource().token;
+  cancellationToken.onCancellationRequested(() => {
+    vscode.window.showInformationMessage("Cancelled server import.");
+  });
+
+  await cmd.get(
+    "reg query " +
+      hkeyLocalMachine +
+      "\\SOFTWARE\\Intersystems\\Cache\\Servers",
+    async (err, data, stderr) => {
+      // Prompt user for username (if not in registry) and password
+      data.split("\n").forEach((serverName) => {
+        // remove HKEY_LOCAL_MACHINE\ and whitespace from the server name
+        const path = serverName.substring(hkeyLocalMachine.length + 1).trim();
+
+        const originalName: string = serverName.split("\\").pop().trim();
+        // Enforce the rules from package.json on the server name
+        const name = originalName.toLowerCase().replace(/[^A-Za-z1-9-]/g, "~");
+        const getProperty = (property: string) =>
+          vsWinReg.GetStringRegKey(hkeyLocalMachine, path, property);
+
+        if (name !== "" && !config.has("servers." + name)) {
+          newServerNames.push(name);
+          let username: string | undefined;
+          try {
+            username = vsWinReg.GetStringRegKey(
+              "HKEY_CURRENT_USER",
+              "Software\\InterSystems\\Cache\\Servers\\" + originalName,
+              "Server User Name",
+            );
+          } catch (e) {
+            // No-op; will assume the key did not exist (it is valid for it not to).
+          }
+
+          if (username === undefined || username === "") {
+            serversMissingUsernames.push(name);
+          }
+
+          const usesHttps = getProperty("HTTPS") === "1";
+          serverDefinitions[name] = {
+            description: getProperty("Comment"),
+            username: (username === undefined || username === "" ? undefined : username),
+            webServer: {
+              host: getProperty("Address"),
+              port: parseInt(getProperty("WebServerPort"), 10),
+              scheme: usesHttps ? "https" : "http",
```

Comment:
Can you also get any `pathPrefix` setting that may exist in the registry entry?

Reply:
@gjsjohnmurray that should map to "CSP Server Instance" I think - right?

Reply:
That's my understanding.